core/trino-main/src/main/java/io/trino/operator/aggregation/state/LongDecimalWithOverflowAndLongStateFactory.java

```diff
@@ -66,27 +79,80 @@ public void addLong(long value)
             longs.add(getGroupId(), value);
         }
 
+        @Override
+        public long[] getDecimalArray()
+        {
+            return unscaledDecimals.getSegment(getGroupId() * 2);
+        }
+
+        @Override
+        public int getDecimalArrayOffset()
+        {
+            return unscaledDecimals.getOffset(getGroupId() * 2);
+        }
+
+        @Override
+        public long getOverflow()
+        {
+            if (overflows == null) {
+                return 0;
+            }
+            return overflows.get(getGroupId());
+        }
+
+        @Override
+        public void setOverflow(long overflow)
+        {
+            // setOverflow(0) must overwrite any existing overflow value
+            if (overflow == 0 && overflows == null) {
+                return;
+            }
+            long groupId = getGroupId();
+            if (overflows == null) {
+                overflows = new LongBigArray();
+                overflows.ensureCapacity(longs.getCapacity());
+            }
+            overflows.set(groupId, overflow);
+        }
+
+        @Override
+        public void addOverflow(long overflow)
+        {
+            if (overflow != 0) {
+                long groupId = getGroupId();
+                if (overflows == null) {
+                    overflows = new LongBigArray();
+                    overflows.ensureCapacity(longs.getCapacity());
+                }
+                overflows.add(groupId, overflow);
+            }
+        }
+
         @Override
         public long getEstimatedSize()
         {
-            return INSTANCE_SIZE + longs.sizeOf() + isNotNull.sizeOf() + unscaledDecimals.sizeOf() + (overflows == null ? 0 : overflows.sizeOf());
+            return INSTANCE_SIZE + longs.sizeOf() + unscaledDecimals.sizeOf() + (overflows == null ? 0 : overflows.sizeOf());
         }
     }
 
     public static class SingleLongDecimalWithOverflowAndLongState
-            extends LongDecimalWithOverflowStateFactory.SingleLongDecimalWithOverflowState
             implements LongDecimalWithOverflowAndLongState
     {
         private static final int INSTANCE_SIZE = instanceSize(SingleLongDecimalWithOverflowAndLongState.class);
+        private static final int SIZE = (int) sizeOf(new long[2]);
 
-        protected long longValue;
+        private final long[] unscaledDecimal = new long[2];
+        private long longValue;
+        private long overflow;
 
         public SingleLongDecimalWithOverflowAndLongState() {}
 
         // for copying
-        private SingleLongDecimalWithOverflowAndLongState(long longValue)
+        private SingleLongDecimalWithOverflowAndLongState(long[] unscaledDecimal, long longValue, long overflow)
```

Comment:
I don't understand. Was this a bug? It wasn't delegating to super constructor.

Does this deserve a separate commit/PR?

Reply:
It looked like a mistake in the current code to me but I don't know enough about how it is used to definitively say whether it would've resulted in a bug.
@kasiafi could you please take a look and clarify.